pkg/proc/proc.go

```diff
@@ -147,6 +147,39 @@ func Continue(dbp Process) error {
 				}
 			}
 			return conditionErrors(threads)
+		case curbp.Active && curbp.Tracepoint:
+			if curbp.returnInfo != nil {
+				curthread.Common().returnValues = curbp.Breakpoint.returnInfo.Collect(curthread)
+				return conditionErrors(threads)
+			}
+			// We have hit a function entry breakpoint, set a
+			// breakpoint at the return address and load return
+			// values.
+			selg := dbp.SelectedGoroutine()
+			topframe, retframe, err := topframe(selg, curthread)
+			if err != nil {
+				return err
+			}
+
+			if topframe.Inlined {
+				if err := next(dbp, false, true); err != nil {
+					return err
+				}
+				return Continue(dbp)
+			}
+
+			sameGCond := SameGoroutineCondition(selg)
```

Comment:
I think there is a problem with concurrency here. Two goroutines can call the same function simultaneously (or one goroutine calls a function, then a second goroutine calls the same function before the first goroutine has returned from it). In that case only the return value from the first function would be output.

Removing this condition doesn't work either, the CALL that entered the function call could be a dynamic dispatch.

IMHO the best to do this is to disassemble every function we want traced and set breakpoints on every `RET` and `CALL runtime.deferreturn` instruction. It's also simpler and takes care of the problem with redefining the meaning of  Breakpoint.Tracepoint and having a recursive call inside Continue.

Reply:
> We would still have a similar concurrency issue

I don't think you would, those can be persistent and don't have any condition associated. Matching between entry/exit can be done in a higher layer (although that's very complicated in general, since some calls never exit because of panics).

> If we allowed multiple conditions for a single breakpoint, we could solve the issue of only one return breakpoint triggering, but we'd still have to be smarter about grouping the entry / exit breakpoints.

You'll need some complicated logic to handle recursive functions: you'd have to keep, for each goroutine, a stack of ReturnBreakpointInfo and keep it sync'd with the target process (removing multiple entries when there's a panic). I'm not even sure it's possible. 

Consider this scenario:
Let's say we are tracing a function f and it's called through a function like this:

```
func dynamicdispatch(fn func() int) {
   fn()
}

dynamicdispatch(f)
```

Let's say there's also a second function g that is also called through dynamicdispatch.

```
- target calls dynamicdispatch(f)
- dynamicdispatch calls f
-> we set a return breakpoint in dynamicdispatch
- f calls some other function
- there's a panic
- target unwinds the stack to well before dynamicdispatch
- time passes
- target calls dynamicdispatch(g), it just so happens that the stack depth is exactly the same as before, since we are not tracing g we don't notice
- g returns
-> we are now at the return breakpoint we set before, on the right goroutine, with the right stack depth, so we read the return values of g as if they were the return values of f, since they have the same signature the result even looks plausible.
```

By setting the return breakpoint on the RET/CALL runtime.deferreturn instruction you don't have this problem: the only way to get at them is to go through `f`. To associate entry with exit you can just compare stack depth by reading the FrameOffset.

I understand that using the disassembly is not exactly tasteful, especially now that we're finally getting rid of the disassembly based prologue detector, but this particular case is not as bad as the prologue detector: normally there's only one way functions exit, the RET instructions, go has another way (CALL runtime.deferreturn) but there is no reason to believe that this will change and a third way will appear.